OrbitVulkanLayer/SubmissionTrackerTest.cpp

```diff
@@ -660,6 +660,67 @@ TEST_F(SubmissionTrackerTest, StartCaptureDuringSubmissionWontWriteData) {
   tracker_.CompleteSubmits(device_);
 }
 
+TEST_F(SubmissionTrackerTest, MultipleSubmissionsWontBeOutOfOrder) {
+  ExpectFourNextReadyQuerySlotCalls();
+  EXPECT_CALL(dispatch_table_, GetQueryPoolResults)
+      .WillOnce(Return(mock_get_query_pool_results_function_not_ready_))
+      .WillRepeatedly(Return(mock_get_query_pool_results_function_all_ready_));
+
+  std::vector<uint32_t> actual_reset_slots;
+  auto mock_reset_query_slots = [&actual_reset_slots](VkDevice /*device*/,
+                                                      const std::vector<uint32_t> slots_to_reset) {
+    actual_reset_slots.insert(actual_reset_slots.end(), slots_to_reset.begin(),
+                              slots_to_reset.end());
+  };
+  EXPECT_CALL(timer_query_pool_, ResetQuerySlots).WillRepeatedly(Invoke(mock_reset_query_slots));
+
+  std::vector<orbit_grpc_protos::CaptureEvent> actual_capture_events;
+  auto mock_enqueue_capture_event =
+      [&actual_capture_events](orbit_grpc_protos::CaptureEvent&& capture_event) {
+        actual_capture_events.emplace_back(std::move(capture_event));
+        return true;
+      };
+  EXPECT_CALL(*producer_, EnqueueCaptureEvent)
+      .Times(2)
+      .WillRepeatedly(Invoke(mock_enqueue_capture_event));
+
+  producer_->StartCapture();
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+
+  // Submission #1:
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.MarkCommandBufferEnd(command_buffer_);
+  std::optional<QueueSubmission> queue_submission_optional_1 =
+      tracker_.PersistCommandBuffersOnSubmit(1, &submit_info_);
+  tracker_.PersistDebugMarkersOnSubmit(queue_, 1, &submit_info_, queue_submission_optional_1);
+  // The results are not yet ready at this point in time.
+  tracker_.CompleteSubmits(device_);
+
+  // Submission #2:
+  tracker_.ResetCommandBuffer(command_buffer_);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.MarkCommandBufferEnd(command_buffer_);
+  std::optional<QueueSubmission> queue_submission_optional_2 =
+      tracker_.PersistCommandBuffersOnSubmit(1, &submit_info_);
+  tracker_.PersistDebugMarkersOnSubmit(queue_, 1, &submit_info_, queue_submission_optional_2);
+  // Now the results are ready for all submissions.
```

Comment:
What causes all submissions here to be ready compared to the `Submission #1` block?

Reply:
Line 665:   
```
EXPECT_CALL(dispatch_table_, GetQueryPoolResults)
      .WillOnce(Return(mock_get_query_pool_results_function_not_ready_))
      .WillRepeatedly(Return(mock_get_query_pool_results_function_all_ready_));
```

In the first submission, we try to get the results for the one command buffer, that is part of the submission, and will note that it is not yet ready. Afterwards, every call will succeed.

Reply:
Ah, I would suggest adding a quick comment for that (either here or where the `EXPECT_CALL` is).

Reply:
Added a comment.